src/Surfnet/StepupGateway/GatewayBundle/Service/GatewayLoginService.php

```diff
@@ -0,0 +1,335 @@
+<?php
+
+/**
+ * Copyright 2018 SURFnet bv
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupGateway\GatewayBundle\Service;
+
+use Exception;
+use SAML2\Constants;
+use SAML2\Response as SAMLResponse;
+use Surfnet\SamlBundle\Entity\IdentityProvider;
+use Surfnet\SamlBundle\Entity\ServiceProvider;
+use Surfnet\SamlBundle\Http\PostBinding;
+use Surfnet\SamlBundle\Monolog\SamlAuthenticationLogger;
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Surfnet\SamlBundle\SAML2\AuthnRequestFactory;
+use Surfnet\SamlBundle\SAML2\ReceivedAuthnRequest;
+use Surfnet\StepupBundle\Service\LoaResolutionService;
+use Surfnet\StepupBundle\Service\SecondFactorTypeService;
+use Surfnet\StepupGateway\GatewayBundle\Exception\RequesterFailureException;
+use Surfnet\StepupGateway\GatewayBundle\Exception\ResponseFailureException;
+use Surfnet\StepupGateway\GatewayBundle\Saml\AssertionAdapter;
+use Surfnet\StepupGateway\GatewayBundle\Saml\Exception\UnknownInResponseToException;
+use Surfnet\StepupGateway\GatewayBundle\Saml\Proxy\ProxyStateHandler;
+use Surfnet\StepupGateway\GatewayBundle\Saml\ResponseBuilder;
+use Surfnet\StepupGateway\GatewayBundle\Saml\ResponseContext;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Entry point for the Stepup login flow.
+ *
+ * See docs/GatewayState.md for a high-level diagram on how this controller
+ * interacts with outside actors and other parts of Stepup.
+ */
+class GatewayLoginService
```

Comment:
I'm not sold of the name either. I've moved the services up to the a seperate folder to keep the gateway services on a different level and renamed the service to LoginService. personally I don't want to split up to much logic now because the main goal was to split up the controller logic to make it testable and not to decouple everything. I know that we could benefit of this in the future and that's much more easier to do now we have these tests.